Make sure temporary file handles are not inherited by child processes

bwijen opened the pull request:
When testing a merge driver which spawns a merge server (for future merges)
I got the following error:
> Rename from 'xxx/.git/index.lock' to 'xxx/.git/index' failed. Should I try again? (y/n)

Only after I stop the merge server the lock is released.
This is caused by [windows handle inheritance](https://msdn.microsoft.com/en-us/library/windows/desktop/ms724466%28v=vs.85%29.aspx)

Starting childs with bInheritHandles==FALSE does not work.
(For some reason the merge commit window did not pop up.)

However creating the handle with bInheritHandle==FALSE does seem to work.
Although I think this should not break anything, it would be great to get some feedback on this.

mcve:
```shell
git init

cat > mergetest.sh << EOF
#!/bin/sh
echo mergetest.sh

sleep 60 &
EOF

git config --local merge.mergetest.driver "./mergetest.sh %O %A %B %P"

echo "*.test merge=mergetest" > .gitattributes
echo first line > fake.test

git add .
git commit -m "initial"

git checkout -b other master
echo other change >> fake.test
git add fake.test
git commit -m "other"

git checkout -b this master
echo this change >> fake.test
git add fake.test
git commit -m "this"

#now try: git merge other
```

bwijen commented:
I reverted my first change, using O_NOINHERIT for all files was indeed to much.
Since git-upload-pack for example, expects inherited handles.

bwijen commented:
@dscho: Any comments on my previous attempt?

dscho commented:
I was a bit busy trying to set up some automatic testing. It's still too slow (3.5h for any single run) but at least it showed that your branch passes the test suite: https://gitlab.com/git-for-windows1/git/commit/9251e5b7bad7db1f6541d4c2ac87894546ee4867/builds

dscho commented on t/t6026-merge-attr.sh:
```diff
@@ -180,4 +180,28 @@ test_expect_success 'up-to-date merge without common ancestor' '
+chmod +x ./custom-merge
+
+test_expect_success 'custom merge backend with delay' '
+	git reset --hard anchor &&
```
Do we need to reset explicitly here? If not, let's skip that statement.

bwijen replied:
This is just to make sure to get into a consistent state. (In case any of the previous tests have failed...)

dscho commented on t/t6026-merge-attr.sh:
```diff
@@ -180,4 +180,28 @@ test_expect_success 'up-to-date merge without common ancestor' '
+test_expect_success 'custom merge backend with delay' '
+	git reset --hard anchor &&
+	echo "* merge=union" >.gitattributes &&
+	echo "text merge=custom" >>.gitattributes &&
```
Shorter: `printf "* merge=union\ntext merge=sleep-one-second\n" >.gitattributes`.

dscho commented on t/t6026-merge-attr.sh:
```diff
@@ -180,4 +180,28 @@ test_expect_success 'up-to-date merge without common ancestor' '
+	git config --replace-all \
+	merge.union.driver "true" &&
+	git config --replace-all \
+	merge.union.name "always ours" &&
```
What is the purpose of this block? Is it needed at all?

bwijen replied:
Needed for the other files, see the "custom merge backend" tests.
(I did however removed other unnecessary lines.)

dscho commented on t/t6026-merge-attr.sh:
```diff
@@ -180,4 +180,28 @@ test_expect_success 'up-to-date merge without common ancestor' '
+#$SHELL_PATH
+sleep 1 &
+EOF
+chmod +x ./custom-merge
```
Let's use the `write_script` function instead. See e.g. https://github.com/git/git/blob/v2.8.3/t/t0001-init.sh#L91-L98

As to the name, let's call it `sleep-one-sec.sh`?

dscho replied:
Oh, and let's put it *inside* the `test_expect_success` block.

dscho commented on t/t6026-merge-attr.sh:
```diff
@@ -180,4 +180,28 @@ test_expect_success 'up-to-date merge without common ancestor' '
+	git config --replace-all \
+	merge.custom.driver "./custom-merge %O %A %B 0 %P" &&
+	git config --replace-all \
+	merge.custom.name "custom merge driver for testing" &&
```
The name is not really needed, is it? Let's keep the test short and sweet; the name "custom" of the custom driver (or "sleep-one-second", as I would prefer) is self-explanatory enough.

dscho commented on t/t6026-merge-attr.sh:
```diff
@@ -180,4 +180,28 @@ test_expect_success 'up-to-date merge without common ancestor' '
+	merge.union.name "always ours" &&
+
+	git config --replace-all \
+	merge.custom.driver "./custom-merge %O %A %B 0 %P" &&
```
The arguments are ignored by the script. Why should we pass them? Let's not.

Also, do we have to `--replace-all`? If we go with the `sleep-one-second` name, then

```sh
git config merge.sleep-one-second.driver ./sleep-one-second.sh
```
should be enough.

dscho commented on t/t6026-merge-attr.sh:
```diff
@@ -180,4 +180,28 @@ test_expect_success 'up-to-date merge without common ancestor' '
+EOF
+chmod +x ./custom-merge
+
+test_expect_success 'custom merge backend with delay' '
```
The label "custom merge backend with delay" describes nicely what this is. Just reading the (short) code would tell us the same, though.

A much more interesting description (one that is currently missing) would be: "custom merge does not lock index".

And that latter description would also help the casual reader to figure out why the heck this needs to be tested.

dscho commented:
Interesting approach. I am not really certain that this is safe, but I guess it is correct code. It certainly passes the test suite now which is necessary to give me a little peace of mind.

Could you address my concerns with the test script? After that I will gratefully merge this PR.

bwijen commented:
I changed the test-script as requested.
Also rebased to the latest version.

Furthermore I also tested on a (virtual) linux box, which shows no failed tests...

dscho commented:
Thank you!
